### DISCUSSION TITLE
Issue with SPI on ESP32Wroom32 and CC1101

### DISCUSSION THREAD

**Discussion by jpitz31** [jgromes/RadioLib#953] [Q&A]
Hi all, I am using the example code: CC1101_Transmit_Address.ino.  I have ran a small program to verify what the default SPI pins are, they are as follows:  MISO = 19, CLK = 18, CS = 5 and MOSI = 23.  I have also wired the GDo0 pin on the CC1101 to GPIO pin 4.  I create the radio object as follows:  CC1101 radio new Module(5, 4, RADIOLIB_NC, RADIOLIB_NC);

I have two CC1101 boards and have tested them on my Flipper Zero using the GPIO pins and they both transmit and receive.

When I compile and execute the program I get the error [CC1101] Transmitting packet ... failed, code -5
Looking in the TypeDef.h file I can see that  -5 is the #define RADIOLIB_ERR_TX_TIMEOUT  constant.

I have double checked that my wiring is proper and have tested both CC1101 chips.  I still get the above error.

The pins that are shown for SPI above are also marked on the ESP32 pin diagram as VSPI.  So I added the below code as a measure to make sure the SPI pins are being read properly.

I have added the #include <SPI.h> at the top of the program and inside of the setup() function near the top I have added the following:

SPIClass SPI1(VSPI);
SPI1.begin();

Now when I compile and execute the code I get back the following error:

[CC1101] Transmitting packet ... failed, code -16

Looking in the TypeDef.h file this error related to the following constant:
RADIOLIB_ERR_SPI_WRITE_FAILED 

So my question is could the SPI wiring be incorrect, or could this be an issue?  I saw an earlier post that has been closed indicating that RadioLib uses the default SPI settings that Arduino sets up.

Also is there anything else I can do to troubleshoot this issue?

Thanks in advance

Joe

**Comment by jgromes** [comment]
> CC1101 radio new Module(5, 4, RADIOLIB_NC, RADIOLIB_NC);

So your GDO2 pin is unconnected? Then the TX timeout error is not very surprising, as GDO2 is used to detect transmission end.

Since the program only errored there, I think SPI issues can be ruled out - it seems like later on you did not initialize it properly, leading to the -16 error. Details about configuring non-default SPI can be found [here](https://github.com/jgromes/RadioLib/wiki/Basics#non-standard-spi-setup), the major points are that you have to call `SPI1.begin` before calling `radio.begin`, and you have to provide the instance of the `SPIClass` you want to use to the `Module` constructor - otherwise it has no way to know you want to use something other than the default.

**Comment by jpitz31** [comment]
@jgromes thanks for the quick response.  I am new the RF so this was great info.  I put the GDO2 in in place and removed the SPI object and now the radio is transmitting.  Thanks so much.  I have tested several CC0010 libraries and find that Radiolib has much better error handling than the others.  Also much easier to setup than RFQuack. 

Thanks again

Joe

**Reply by jgromes** [reply]
No worries, glad you were able to get it running - feel free to ask more questions if any come up ;)
